macros/src/lib.rs

```diff
@@ -277,3 +280,140 @@ pub fn modal(input: TokenStream) -> TokenStream {
+    let mut command_idents = vec![];
+
+    // collect each ImplItem in a stream
+    let mut impl_body = quote!();
+
+    for item in item_impl.items.iter() {
+        // if it's a function...
+        if let syn::ImplItem::Fn(f) = item {
+            // ... and it's a command
+            if let Some(attr) = f.attrs.iter().find(|attr| is_command_attr(attr)) {
+                // add to command list
+                command_idents.push(f.sig.ident.clone());
+
+                // Turn a syn::Attribute into command::CommandArgs
+                let attr_args = match attr.meta.require_list() {
+                    Ok(x) => &x.tokens,
+                    Err(e) => return e.into_compile_error().into(),
+                };
+
+                let command_args =
+                    match darling::ast::NestedMeta::parse_meta_list(quote! {#attr_args}) {
+                        Ok(x) => x,
+                        Err(e) => {
+                            return e.into_compile_error().into();
+                        }
+                    };
+                let command_args =
+                    match <command::CommandArgs as darling::FromMeta>::from_list(&command_args) {
+                        Ok(x) => x,
+                        Err(e) => return e.write_errors().into(),
+                    };
+
+                let new_args = command_args.from_group_args(&group_args);
+                let function = syn::ItemFn {
+                    attrs: vec![],
+                    vis: f.vis.clone(),
+                    sig: f.sig.clone(),
+                    block: Box::new(f.block.clone()),
+                };
+                let cmd: proc_macro2::TokenStream = match command::command(new_args, function) {
+                    Ok(x) => x.into(),
+                    Err(e) => e.write_errors().into(),
+                };
+
+                impl_body = quote!(
+                    #impl_body
+                    #cmd
+                );
+                // repeated because if let chains are unstable <https://github.com/rust-lang/rust/issues/53667>
+            } else {
+                impl_body = quote!(
+                    #impl_body
+                    #item
+                );
+            };
+        } else {
+            impl_body = quote!(
+                #impl_body
+                #item
+            );
+        }
+    }
+
+    quote! {
+        impl #name {
+            fn commands() -> Vec<Command<Data, Error>> {
+                vec![#(#name::#command_idents()),*]
+            }
+
+            #impl_body
+        }
+    }
+    .into()
+}
+
+/**
+Returns true if an `Attribute` has `path` equal to "poise::command" or "command"
+*/
+fn is_command_attr(attr: &syn::Attribute) -> bool {
+    let path = attr
+        .path()
+        .segments
+        .iter()
+        .map(|s| format!("{}", s.ident))
+        .collect::<Vec<String>>()
+        .join("::");
+
+    path == "poise::command" || path == "command"
+}
```

Comment:
Can't this be implemented without the formatting and the vector allocation and stuff? Just check the first and second segments?

Reply:
This _is_ ugly.
What do you think of
```rust
if let Some(segment) = path.attr.path().segments.last() {
    segment.ident == command
} else {
    false
}
```

Reply:
That wouldn't handle the case of `bogus::command` correctly

Reply:
Neither would the previous version, tbf, the user could just
```rust
use poise::command;
// use bogus::command;

#[command(...)]
```

And, afaik, a macro can't know which is which.
I don't think it is s problem though, since unless `bogus::command` takes the same attributes as `poise::command`.

You could always only match `poise::command`, I guess.